guides/rails_guides/epub.rb

```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/env ruby
+# frozen_string_literal: true
+
+require "nokogiri"
+require "fileutils"
+require "yaml"
+require "date"
+
+require "rails_guides/epub_packer"
+
+module Epub
+  extend self
+
+  def generate(output_dir, epub_outfile)
+    book_dir = File.absolute_path(output_dir)
+    output_dir = File.absolute_path(File.join(output_dir, ".."))
+    Dir.chdir book_dir do
+      puts "=> Using book dir: #{book_dir}"
+      puts "=> Fixing filenames in Table of Contents"
+      fix_file_names
+    end
+    Dir.chdir output_dir do
+      puts "=> Using output dir: #{output_dir}"
+      puts "=> Generating meta files"
+      generate_meta_files
+
+      puts "=> Generating EPUB"
+
+      EpubPacker.pack("./", epub_outfile)
+
+      puts "=> Done Generating EPUB"
+    end
+  end
+
+  def open_toc_doc(toc)
+    Nokogiri::XML(toc).xpath("//ncx:content", "ncx" => "http://www.daisy.org/z3986/2005/ncx/")
+  end
+
+  def generate_meta_files
+    FileUtils.mkdir_p("META-INF")
+    File.write("META-INF/container.xml", <<~CONTENT)
+    <?xml version="1.0" encoding="UTF-8"?>
+    <container xmlns="urn:oasis:names:tc:opendocument:xmlns:container" version="1.0">
+      <rootfiles>
+          <rootfile full-path="OEBPS/rails_guides.opf" media-type="application/oebps-package+xml"/>
+      </rootfiles>
+    </container>
+    CONTENT
+  end
+
+  def is_name_invalid(name)
+    (/[0-9]/ === name.chars.first)
+  end
+
+  def fix_file_names
+    # opf file: item->id and itemref->idref attributes does not support values starting with a number
+    toc = File.read("toc.ncx")
+    toc_html = File.read("toc.html")
+    opf = File.read("rails_guides.opf")
+
+    doc = open_toc_doc(toc)
+    doc.each { |c|
```

Comment:
Style nit: We prefer `do/end` blocks over braces when blocks are multi-line

Reply:
Fixed